### PR TITLE
Enable debug and coverage.

### PR DESCRIPTION
**Pull request by kobrineli**
Hi! I've added `ENABLE_DEBUG` and `ENABLE_COV` variables to enable building with them

### CODE REVIEW

**Review by SweetVishnya — APPROVED** [review]

**Review by ligurio — CHANGES_REQUESTED** [review]
Thanks for the patch! See my comments inline.

File: cmake/BuildLua.cmake [hunk 1]
```diff
@@ -7,16 +7,27 @@ macro(build_lua LUA_VERSION)
     set(CFLAGS "-DLUAI_ASSERT=1 -DLUA_USE_APICHECK=1 -fsanitize=fuzzer-no-link")
     set(LDFLAGS "-fsanitize=fuzzer-no-link")
 
+    if (ENABLE_DEBUG)
```
[the remark this answers is not in the document]
Comment (kobrineli) [reply cmake/BuildLua.cmake]: I thought that cmake adds `-g` automatically as well, but I didn't find debug info, so the check will not be redundant.

Remove `ENABLE_DEBUG` var, added usage of `CMAKE_BUILD_TYPE`

File: cmake/BuildLua.cmake [hunk 2]
```diff
@@ -7,16 +7,27 @@ macro(build_lua LUA_VERSION)
...
     endif (ENABLE_ASAN)
 
     if (ENABLE_UBSAN)
-        set(CFLAGS "${CFLAGS} -fsanitize=fuzzer-no-link,undefined")
-        set(LDFLAGS "${LDFLAGS} -fsanitize=fuzzer-no-link,undefined")
+        set(CFLAGS "${CFLAGS} -fsanitize=undefined")
+        set(LDFLAGS "${LDFLAGS} -fsanitize=undefined")
     endif (ENABLE_UBSAN)
 
+    if (ENABLE_COV)
```
[the remark this answers is not in the document]
Comment (kobrineli) [reply cmake/BuildLua.cmake:25]: done

File: cmake/BuildLua.cmake (hunk 1 above)
[the remark this answers is not in the document]
Comment (ligurio) [reply cmake/BuildLua.cmake]: CMakeLists.txt:
```
add_executable(sample sample.c)
```

sample.c:
```c
int main() {
        return 0;
}
```

```
cmake -S . -B build -DCMAKE_BUILD_TYPE=Debug
<snipped>
-- The C compiler identification is GNU 11.3.0
-- The CXX compiler identification is GNU 11.3.0                      
-- Detecting C compiler ABI info                                                                                                            
-- Detecting C compiler ABI info - done                                                                                                     
-- Check for working C compiler: /bin/cc - skipped                                                                                          
-- Detecting C compile features                                                                                                             
-- Detecting C compile features - done                                                                                                      
-- Detecting CXX compiler ABI info                                    
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Configuring done
-- Generating done
-- Build files have been written to: /home/sergeyb/sources/cmake-sample/build
```



Build log:

```
[0] ~/sources/cmake-sample/$ cd build
[0] ~/sources/cmake-sample/build$ make VERBOSE=1
/usr/bin/cmake -S/home/sergeyb/sources/cmake-sample -B/home/sergeyb/sources/cmake-sample/build --check-build-system CMakeFiles/Makefile.cmake 0
/usr/bin/cmake -E cmake_progress_start /home/sergeyb/sources/cmake-sample/build/CMakeFiles /home/sergeyb/sources/cmake-sample/build//CMakeFiles/progress.marks
make  -f CMakeFiles/Makefile2 all
make[1]: Entering directory '/home/sergeyb/sources/cmake-sample/build'
make  -f CMakeFiles/sample.dir/build.make CMakeFiles/sample.dir/depend
make[2]: Entering directory '/home/sergeyb/sources/cmake-sample/build'
cd /home/sergeyb/sources/cmake-sample/build && /usr/bin/cmake -E cmake_depends "Unix Makefiles" /home/sergeyb/sources/cmake-sample /home/sergeyb/sources/cmake-sample /home/sergeyb/sources/cmake-sample/build /home/sergeyb/sources/cmake-sample/build /home/sergeyb/sources/cmake-sample/build/CMakeFiles/sample.dir/DependInfo.cmake --color=
make[2]: Leaving directory '/home/sergeyb/sources/cmake-sample/build'
make  -f CMakeFiles/sample.dir/build.make CMakeFiles/sample.dir/build
make[2]: Entering directory '/home/sergeyb/sources/cmake-sample/build'
[ 50%] Building C object CMakeFiles/sample.dir/sample.c.o
/bin/cc   -g -MD -MT CMakeFiles/sample.dir/sample.c.o -MF CMakeFiles/sample.dir/sample.c.o.d -o CMakeFiles/sample.dir/sample.c.o -c /home/sergeyb/sources/cmake-sample/sample.c
[100%] Linking C executable sample
/usr/bin/cmake -E cmake_link_script CMakeFiles/sample.dir/link.txt --verbose=1
/bin/cc -g -rdynamic CMakeFiles/sample.dir/sample.c.o -o sample 
make[2]: Leaving directory '/home/sergeyb/sources/cmake-sample/build'
[100%] Built target sample
make[1]: Leaving directory '/home/sergeyb/sources/cmake-sample/build'
/usr/bin/cmake -E cmake_progress_start /home/sergeyb/sources/cmake-sample/build/CMakeFiles 0
```

compilation flags for sample already includes `-g`: `/bin/cc -g -rdynamic CMakeFiles/sample.dir/sample.c.o -o sample `

[the remark this answers is not in the document]
Comment (kobrineli) [reply cmake/BuildLua.cmake]: @ligurio 
I agree

But in the lua external project in the build system you use the explicit build command `BUILD_COMMAND cd <SOURCE_DIR> && make -j CC=${CMAKE_C_COMPILER} CFLAGS=${CFLAGS} LDFLAGS=${LDFLAGS}`, which use only flags for lib build, which were explicitly declared earlier, so we need to add `-g` to `CFLAGS` by ourselves

[the remark this answers is not in the document]
Comment (ligurio) [reply cmake/BuildLua.cmake]: Aaah, got it!

I propose to replace `-g` with `CMAKE_C_FLAGS_DEBUG` (there is also a `CMAKE_CXX_FLAGS_DEBUG`, but Lua is a pure C project).

[the remark this answers is not in the document]
Comment (kobrineli) [reply cmake/BuildLua.cmake]: Looks good, agree

Fixed

[the remark this answers is not in the document]
Comment (kobrineli) [reply cmake/BuildLua.cmake]: @ligurio 

Can I resolve the conversation then?

**Review by ligurio — CHANGES_REQUESTED** [review]

File: cmake/BuildLua.cmake (hunk 1 above)
[the remark this answers is not in the document]
Comment (ligurio) [reply cmake/BuildLua.cmake]: Sure

File: CMakeLists.txt [hunk 3]
```diff
@@ -9,6 +9,7 @@ option(USE_LUA "Use PUC Rio Lua library" OFF)
 option(USE_LUAJIT "Use LuaJIT library" OFF)
 option(ENABLE_ASAN "Enable AddressSanitizer" OFF)
 option(ENABLE_UBSAN "Enable UndefinedBehaviorSanitizer" OFF)
+option(ENABLE_COV "Enable coverage instrumentation" OFF)
```
[the remark this answers is not in the document]
Comment (kobrineli) [reply CMakeLists.txt:12]: @ligurio 
fixed

**Pull request merged**